C++ API parity: MarginRankingLoss

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28783 C++ API parity: BCEWithLogitsLoss
* #28918 C++ API parity: isfinite
* **#29000 C++ API parity: MarginRankingLoss**

Differential Revision: [D18271855](https://our.internmc.facebook.com/intern/diff/D18271855)

pbelevich committed:
C++ API parity: MarginRankingLoss

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MarginRankingLoss"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MarginRankingLoss"


Differential Revision: [D18271855](https://our.internmc.facebook.com/intern/diff/D18271855)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MarginRankingLoss"


Differential Revision: [D18271855](https://our.internmc.facebook.com/intern/diff/D18271855)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MarginRankingLoss"


Differential Revision: [D18271855](https://our.internmc.facebook.com/intern/diff/D18271855)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MarginRankingLoss"


Differential Revision: [D18271855](https://our.internmc.facebook.com/intern/diff/D18271855)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/options/loss.h:
```diff
@@ -223,5 +223,17 @@ struct TORCH_API PoissonNLLLossOptions {
+  typedef c10::variant<enumtype::kNone, enumtype::kMean, enumtype::kSum> reduction_t;
+
+  /// Has a default value of `0`.
+  TORCH_ARG(float, margin) = 0;
```
I believe we should use
```suggestion
  TORCH_ARG(double, margin) = 0;
```
because `float` in Python is always 64-bit (there are a few other places in the C++ API where we use `TORCH_ARG(float, ...)`, and we will need to change them as well :/ (Not part of this PR))

yf225 reviewed (commented):
@pbelevich Thanks so much for the awesome work! I have one very minor comment

pbelevich committed:
Update on "C++ API parity: MarginRankingLoss"


Differential Revision: [D18271855](https://our.internmc.facebook.com/intern/diff/D18271855)

[ghstack-poisoned]
